lib/Driver/Compilation.cpp

```diff
@@ -714,45 +721,81 @@ namespace driver {
+                                 llvm::SetVector<const Job *> &Set) {
+      if (Set.empty())
         return;
       if (Comp.ShowJobLifecycle)
         llvm::outs() << "Forming batch job from "
-                     << CurrentBatch.size() << " constituents\n";
+                     << Set.size() << " constituents\n";
       auto const &TC = Comp.getToolChain();
-      auto J = TC.constructBatchJob(CurrentBatch.getArrayRef(), Comp);
+      auto J = TC.constructBatchJob(Set.getArrayRef(), Comp);
       if (J)
         Batches.insert(Comp.addJob(std::move(J)));
-      CurrentBatch.clear();
+      Set.clear();
     }
 
-    /// Return true iff \p Cmd can be expanded by \p CurrentBatch, meaning
-    /// that \p CurrentBatch is smaller than \p TargetBatchSize and \p Cmd
-    /// is batch-combinable with the equivalence class of \p CurrentBatch
-    /// (as represented by element 0 of \p CurrentBatch).
-    bool canExpandBatch(const Job *Cmd,
-                        llvm::SetVector<const Job *> &CurrentBatch,
-                        size_t TargetBatchSize) {
-      auto const &TC = Comp.getToolChain();
-      return (CurrentBatch.empty() ||
-              (TC.jobsAreBatchCombinable(Comp, Cmd, CurrentBatch[0]) &&
-               CurrentBatch.size() < TargetBatchSize));
+    /// Given 0 <= \p i < \p P.size(), advance i to the "next" partition,
+    /// which means filling each partition in order (distributing
+    /// any division-remainder evenly between partitions).
+    void nextPartition(size_t &i,
+                       BatchPartition const &Partition,
+                       size_t NumJobs) {
+      assert(i < Partition.size());
+      size_t Remainder = NumJobs % Partition.size();
+      size_t TargetSize = NumJobs / Partition.size();
+      // Spread remainder evenly across partitions by adding 1 to the target
+      // size of the first Remainder of them.
+      if (i < Remainder)
+        TargetSize++;
+      if (Partition[i].size() >= TargetSize)
+        ++i;
+      assert(i < Partition.size());
     }
 
-    /// If \p CurrentBatch can't be expanded with \p Cmd, form a new \c BatchJob
-    /// from \p CurrentBatch, add it to \p Batches, and reset\p CurrentBatch;
-    /// then in either case, insert \p Cmd into \p CurrentBatch.
-    void expandBatch(const Job *Cmd,
-                     CommandSet &Batches,
-                     llvm::SetVector<const Job *> &CurrentBatch,
-                     size_t TargetBatchSize) {
-      if (!canExpandBatch(Cmd, CurrentBatch, TargetBatchSize)) {
-        formBatchJobFromCurrentBatch(Batches, CurrentBatch);
+    /// Shuffle \p Batchable if -driver-batch-seed is nonzero.
+    void maybeShuffleBatchable(CommandSetVector &Batchable) {
+      if (Comp.BatchSeed != 0) {
+        std::minstd_rand gen(Comp.BatchSeed);
+        auto tmp = Batchable.takeVector();
+        std::shuffle(tmp.begin(), tmp.end(), gen);
+        Batchable = CommandSetVector(tmp.begin(), tmp.end());
```

Comment:
Nitpick: maybe this function should just return a `std::vector` (and accept `Batchable` by value, and `partitionIntoBatches` can do the same) so as not to pay the cost of re-uniqueing the elements. It's only used as an array from here on out.

Reply:
Oh, yeah, actually there are a bunch of needless setvector cases after the initial split that should just be vectors. Good catch, fixing.